### ISSUE TITLE
Deployment (Heroku button)

### ISSUE THREAD

**Issue by olimart** [stripe/stripe-mock#42]
Hi,
this is a suggestion but it would be great to be able to deploy an instance through the click of a button 

https://devcenter.heroku.com/articles/heroku-button

Thanks.

**Comment by brandur-stripe** [comment]
Yeah, this could be a good idea. As far as apps go to deploy, this should also be _really_ easy even without a button. There's not much in the way of dependencies, so a clone, `heroku create`, and `git push heroku master` should be enough to do it.

It also seems to be a little bit wasteful to have separate Heroku apps for every user when they're all doing the same thing — we should probably have a hosted version of this somewhere.

**Comment by olimart** [comment]
> It also seems to be a little bit wasteful to have separate Heroku apps for every user when they're all doing the same thing

Indeed.

I'll try to clone the app and deploy. Thanks

**Comment by dja** [comment]
Just tried this and couldn't get it to run on Heroku with a simple create then deploy. What's the work involved to create a hosted version of this?

**Comment by brandur-stripe** [comment]
> Just tried this and couldn't get it to run on Heroku with a simple create then deploy. What's the work involved to create a hosted version of this?

This should be quite trivial to get running on Heroku — the project's distributed as a single binary and doesn't use any kind of exotic build trickery. Did it look like the Go buildpack was detected properly? Also, Heroku used to like that certain Go version annotations were set in a `vendor.json` file — did you get an error about that?

**Comment by shipstar** [comment]
I'm not sure if my issue was the same as @dja's, but when I did a simple Heroku create / deploy, the app hung indefinitely. The buildpack was detected and I _think_ the app was running, just on the wrong port. Heroku uses a `PORT` env var instead of the `--port` flag used by this app, so I think their networking just didn't have a way to detect the 12111 default port.

I've never written go before, but I added the following code to `getListener` in `main.go`:

```go
if port == 0 {
    // added the next two lines
    port, err = strconv.Atoi(os.Getenv("PORT"))
    if err != nil {
        port = defaultPort
    }
}
```

Once I committed that and redeployed, it started working for me.

I'll let someone who's more familiar with Go and/or this project decide if this warrants a PR, but I hope it helps!

**Comment by ob-stripe** [comment]
@shipstar You should be able to set up your Heroku Procfile to do something like this:
```
web: stripe-mock --port $PORT
```

**Comment by brandur-stripe** [comment]
> @shipstar You should be able to set up your Heroku Procfile to do something like this:

+1.

That said, it should probably also respect the value of `PORT` since that's become quite a common convention. I have to take a look at `stripe-mock` a little later today, and I'll put a patch together for that.

**Comment by brandur-stripe** [comment]
Released a fix for the `PORT` problem in v0.11.0.
